samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java

```diff
@@ -69,6 +71,7 @@ public ZkLeaderElector(String processorIdStr, ZkUtils zkUtils) {
   @VisibleForTesting
   public ZkLeaderElector(String processorIdStr,
                          ZkUtils zkUtils,
+                         @Nonnull
                          IZkDataListener previousProcessorChangeListener) {
     this.processorIdStr = processorIdStr;
```

Comment:
Do we have any other callers for this method? Given you have modified the test caller to use different constructor? I want to make sure this non-null change is necessary.

Reply:
Another way to do this to add a null check, and call `ZkLeaderElector(String processorIdStr ZkUtils zkUtils)` if null.

There are no internal callers that pass in null here, except our unit tests.

Reply:
https://developer.android.com/reference/androidx/annotation/VisibleForTesting

Confirm this should be safe. This annotation would make the method `private` by default, so it is only used by our tests.

Reply:
The annotation doesn't do much and is mostly for readability. Moreover, the annotation used in this from guava and I highly doubt anything is enforced through the annotation.

That said, if there are no usages in the code, you can remove this constructor. I am leaning towards making no change and have the tests use the same constructor since the behavior of two constructors differ slightly with validation on zkUtils.

Reply:
@mynameborat
The path forward isn't clear to me.

1. We want to remove the `@NonNull` line
2. Do we want to remove this constructor?
3. Need to choose one of the following:
     a. Pass in a non-null ZkDataListener in the tests, or
     b. Use the constructor without the last argument.

Reply:
Without this change, the unit test just fail with NPE.